### PR TITLE
CORTX-31923: ListObjectVersions and DeleteObject with versioning RBF

### PR DESCRIPTION
**Pull request by apoorva-rao**
Signed-off-by: Apoorva Rao <apoorva.d.rao@seagate.com>

# Problem Statement
- Fixes for ListObjectVersions and DeleteObject (with versionin support) tests

# Design
-  https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/980353025/F-20F+S3+Versioning+for+Simple+Object+Operations

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [x] New/Affected tests are executed on Latest Build
-  [x] Attach test execution logs
-  [x] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [x] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: libs/s3/s3_versioning_common_test_lib.py [hunk 1]
```diff
@@ -421,12 +424,13 @@ def upload_version(s3_test_obj: S3TestLib, bucket_name: str, object_name: str,
     versions_dict[object_name]["is_latest"] = version_id
 
 
-def delete_version(s3_ver_test_obj: S3VersioningTestLib, bucket_name: str, object_name: str,
-                   versions_dict: dict, version_id: str = None,
+def delete_version(s3_test_obj: S3TestLib, s3_ver_test_obj: S3VersioningTestLib, bucket_name: str,
+                   object_name: str, versions_dict: dict, version_id: str = None,
```
Comment (DhobaleVishal) [inline libs/s3/s3_versioning_common_test_lib.py:428]: If this is versioning specific method then version_id should be compulsory parameter.

File: libs/s3/s3_test_lib.py [hunk 2]
```diff
@@ -640,19 +640,24 @@ def get_object(
             bucket: str = None,
             key: str = None,
             ranges: str = None,
-            raise_exec: bool = True) -> tuple:
+            raise_exec: bool = True,
```
Comment (DhobaleVishal) [inline libs/s3/s3_test_lib.py:643]: Better to keep this optional parameters in kwargs.

File: libs/s3/s3_test_lib.py [hunk 3]
```diff
@@ -640,19 +640,24 @@ def get_object(
...
         :param key: Key of the object to get.
         :param ranges: Byte range to be retrieved
         :param bucket: The bucket name containing the object.
+        :param skip_polling: Skip retry for GET Object, in case of failures
         :return: (Boolean, Response)
         """
         try:
             LOGGER.info("Retrieving object from a bucket")
-            response = poll(super().get_object, bucket, key, ranges)
+            if not skip_polling:
```
Comment (DhobaleVishal) [inline libs/s3/s3_test_lib.py:657]: Is this expected in RGW?

**Review by DhobaleVishal — CHANGES_REQUESTED** [review]

File: libs/s3/s3_test_lib.py (hunk 3 above)
Comment (apoorva-rao) [reply libs/s3/s3_test_lib.py:657]: Polling was causing issues in test code which ends up waiting for failure while the lib function is polling for success unnecessarily, so added an option to bypass it

**Review by DhobaleVishal — CHANGES_REQUESTED** [review]

File: libs/s3/s3_test_lib.py (hunk 2 above)
Comment (apoorva-rao) [reply libs/s3/s3_test_lib.py]: done

File: libs/s3/s3_versioning.py [hunk 4]
```diff
@@ -65,7 +65,10 @@ def list_object_versions(self, bucket_name: str = None, optional_params: dict =
             Delimiter, EncodingType, KeyMarker, MaxKeys, Prefix, VersionIdMarker
         :return: response
         """
-        response = self.s3_client.list_object_versions(Bucket=bucket_name, **optional_params)
+        if optional_params is not None:
```
[the remark this answers is not in the document]
Comment (apoorva-rao) [reply libs/s3/s3_versioning.py:68]: the current check is more explicit, so we can retain this as-is

**Review by DhobaleVishal — APPROVED** [review]

File: libs/s3/s3_versioning_common_test_lib.py [hunk 5]
```diff
@@ -364,16 +366,17 @@ def upload_versions(s3_test_obj: S3TestLib, s3_ver_test_obj: S3VersioningTestLib
...
+    if obj_list:
+        for versioning_config, object_name, count in obj_list:
+            resp = s3_ver_test_obj.put_bucket_versioning(bucket_name=bucket_name,
+                                                        status=versioning_config)
+            assert_utils.assert_true(resp[0], resp[1])
+            for _ in range(count):
+                file_path = SystemRandom().choice(file_paths)  # nosec
+                chk_null_version = False if versioning_config == "Enabled" else True
+                upload_version(s3_test_obj, bucket_name=bucket_name, file_path=file_path,
+                            object_name=object_name, versions_dict=versions,
```
Comment (sonalk0209) [inline libs/s3/s3_versioning_common_test_lib.py:378]: please check indentation on 378 and 379

**Review by sonalk0209 — APPROVED** [review]

File: libs/s3/s3_versioning_common_test_lib.py (hunk 5 above)
Comment (apoorva-rao) [reply libs/s3/s3_versioning_common_test_lib.py]: Addressed in latest commit

**Pull request merged**